### PR TITLE
Separate versioned Android ABIs

### PR DESCRIPTION
**Pull request by sjchmiela**
# Why

`expoview` AARs at the moment contain multiple versions of Expo SDK (`expoview@30.0.0` contains ABI 30 and back to 24). Another inconvenience of the current setup is that it is not possible to strip out universal modules from Android apps.

We would like to be able to:
- build apps with only a single Expo SDK
- not have to bundle unimodules into AARs, rather include them as real code in `expoview-abiXX_X_X` libs 

# How

- instead of packaging all old ABIs in a single `expoview` library, I split `expoview` into `expoview-abi{30,29,28,27,26,25,24}_0_0`  ➡️ this will let us include only one of those in standalone applications
- every `expoview-abiXX_0_0` depends on `expoview`
- `expoview` contains both `versioned.`, i. e. unversioned SDK code and the `host.`, kernel code,
- references to unversioned code is formatted with `// BEGIN_SDK_XX` comments` that will be properly removed by XDL (see https://github.com/expo/expo-cli/pull/71),
- from now on I think it would be ok to build new standalone apps as we do on iOS, i. e. let the unversioned code be used as an ABI code.
- `tools` have been updated to cut new releases the new way,
- we could rethink whether distributing `expoview` as an editable code library (linked like a normal RN library, not an AAR) would be ok
- in new Turtle/detached apps we'll include only unversioned code and set `Constants.TEMPORARY_ABI_VERSION` to proper ABI version.

# Test plan

- [x] Turtle with single SDK feature has been deployed to staging and works properly, see https://github.com/expo/universe/pull/3215,
- [x] creating a new SDK version for use in Expo Client following [this new section in README](https://github.com/expo/expo/pull/2330/files#diff-07b723ae57aaa00bce72a3022cf09ecd) works ok, NCL with `app.json.sdkVersion = 31` runs normally,
- [x] revisit `build.gradle`s and `AndroidManifest.xml`s of new `expoview-abiXX_X_X` subprojects
- [x] releasing AAR with a new SDK works all right.

### CODE REVIEW

File: android/app/build.gradle [hunk 1]
```diff
@@ -180,12 +180,71 @@ dependencies {
...
+  // BEGIN_SDK_29
+  implementation(project(':expoview-abi29_0_0')) {
+    exclude group: 'com.squareup.okhttp3', module: 'okhttp'
+    exclude group: 'com.squareup.okhttp3', module: 'okhttp-urlconnection'
+  }
+  // END_SDK_29
+
+  // BEGIN_SDK_28
+  implementation(project(':expoview-abi28_0_0')) {
+    exclude group: 'com.squareup.okhttp3', module: 'okhttp'
```
Comment (esamelson) [inline android/app/build.gradle]: tiny nit, just not sure if these excludes are necessary for SDK 28 and below?

File: android/expoview/build.gradle [hunk 2]
```diff
@@ -96,13 +100,16 @@ task generateDynamicMacros(type: Exec) {
...
+// REMOVE_WHEN_VERSIONING_TO_HERE
 
 dependencies {
   api fileTree(dir: 'libs', include: ['*.jar'])
-  api 'com.android.support:multidex:1.0.1'
+  api 'com.android.support:multidex:1.0.3'
 
+  // WHEN_VERSIONING_REPLACE_WITH_DEPENDENCIES
+
+  // REMOVE_WHEN_VERSIONING_FROM_HERE
```
Comment (esamelson) [inline android/expoview/build.gradle]: very tiny nit: can we change this so it matches the syntax of the other comments (`WHEN_VERSIONING_REMOVE_FROM_HERE`)? 😅

File: tools/README.md [hunk 3]
```diff
@@ -38,3 +38,14 @@ If you want to publish a package with specific version:
 ```
 ./publish-packages.sh --version="1.2.3" --scope="expo-permissions"
 ```
+
+## Versioning Android
```
Comment (esamelson) [inline tools/README.md]: could you make sure that the release steps in universe/internal-docs are up-to-date as well?

**Review by esamelson — APPROVED** [review]
This is awesome 🎉 💯 Thanks for taking this on! I don't really have any substantive comments, the approach you've taken seems solid.

Also, I'm definitely open to distributing expoview as an aar instead of a library. Let's land this first and then discuss that. (I think the main thing we want to avoid is making people compile RN itself from source.)

**Comment by aniltirola** [comment]
I like the idea of 'distributing expoview as an editable code library'.
Thanks for your hard work
anil

**Comment by aniltirola** [comment]
ok, then another possibility to help the user slimming down their apks is a small section in documentation how to exclude more modules like you do it in 
`android/app/build.gradle` with
`exclude group: 'com.squareup.okhttp3', module: 'okhttp'`

.... but I don't know if android 'really excludes' when building apk.... :confused: 🤔🤔

**Comment by aniltirola** [comment]
ok, then another possibility to help the user slimming down their apks is a small section in documentation how to exclude more modules like you do it in
android/app/build.gradle with
exclude group: 'com.squareup.okhttp3', module: 'okhttp'

.... but I don't know if android 'really excludes' when building apk.... 😕 🤔🤔

**Pull request merged**